ansible_base/lib/abstract_models/common.py

```diff
@@ -205,3 +206,30 @@ def summary_fields(self):
 
     def __str__(self):
         return self.name
+
+
+class ImmutableModel:
+    # In case the model is extending CommonModel.
+    modified = None
+    modified_by = None
+
+    def __init_subclass__(cls, **kwargs):
+        super().__init_subclass__(**kwargs)
+
+        # MRO dictates that we must be first.
+        if cls.__bases__[0] is not ImmutableModel:
+            raise ValueError(f"ImmutableModel must be the first base class for {cls.__name__}")
+
+    def save(self, *args, **kwargs):
+        if self.pk:
+            raise ValueError(f"{self.__class__.__name__} is immutable and cannot be modified.")
+
+        return super().save(*args, **kwargs)
+
+    def __getattribute__(self, name):
+        # This is a bit of a hack, there is no good way to remove these fields, because they exist up the chain
+        # if the model inherits from CommonModel.
+        if name in ('modified', 'modified_by'):
+            raise FieldDoesNotExist(f"{self.__class__.__name__} has no attribute {name}")
```

Comment:
I don't follow why you need to do this in the first place, and then I don't understand why you would want this exception in the case it happens. The normal behavior is `AttributeError`

```
In [2]: JobTemplate.objects.first().foobar
---------------------------------------------------------------------------
AttributeError                            Traceback (most recent call last)
Cell In[2], line 1
----> 1 JobTemplate.objects.first().foobar

AttributeError: 'JobTemplate' object has no attribute 'foobar'
```

And I don't know how we wouldn't get this for "free" by doing nothing.

Reply:
I spent a bunch of time on this. We don't get it for "free" for a few reasons.

`CommonModel` will set those fields. In the normal case, that means those fields get updated and when you save, they'll get written to the database. But since in this case they aren't fields (they are `None`), they'll end up just getting set as regular instance variables.

Why is this a problem? Well, because it's confusing if you `dir()` an instance of `ImmutableModel, CommonModel` after you `save()` it, and see those fields. And it's even more confusing if they contain an actual value.

So, then why can't we just eliminate the fields? Well, largely because they are class variables.

There is interesting play here:

```python
In [1]: class Foo():
   ...:     x = 39
   ...:

In [2]: foo = Foo()

In [3]: foo.x
Out[3]: 39

In [4]: hasattr(foo, 'x')
Out[4]: True

In [5]: delattr(foo, 'x')
---------------------------------------------------------------------------
AttributeError                            Traceback (most recent call last)
Cell In[5], line 1
----> 1 delattr(foo, 'x')

AttributeError: 'Foo' object has no attribute 'x'
```

Okay, fair enough, you have to delete them from the class.

```python
In [6]: hasattr(Foo, 'x')
Out[6]: True

In [7]: delattr(Foo, 'x')

In [8]: foo.x
---------------------------------------------------------------------------
AttributeError                            Traceback (most recent call last)
Cell In[8], line 1
----> 1 foo.x

AttributeError: 'Foo' object has no attribute 'x'
```

That kind of works. The fields don't need to exist anymore on `ImmutableModel` classes once the classes are created enough for the Django metaclass logic to run and so on, enough to notice that `modified` and `modified_by` _aren't fields_ and _should not_ be included in migrations.

But wait - if we `delattr()` them from `ImmutableModel` and the model being defined inherits from `ImmutableModel, CommonModel`, then MRO dictates that it will just fall back to using the field from `CommonModel` when you try to access one of those fields. But now you're broken, because the migration said those fields don't exist, so now you call `modified_by` and Django tries to resolve an FK relationship that isn't actually there, so you get an error _deep down_ in the Django code:

```
self = <django.db.models.fields.related.ForeignKey: modified_by>

    def resolve_related_fields(self):
        if not self.from_fields or len(self.from_fields) != len(self.to_fields):
            raise ValueError(
                "Foreign Object from and to fields must be the same non-zero length"
            )
        if isinstance(self.remote_field.model, str):
>           raise ValueError(
                "Related model %r cannot be resolved" % self.remote_field.model
            )
E           ValueError: Related model 'test_app.User' cannot be resolved

.tox/311/lib/python3.11/site-packages/django/db/models/fields/related.py:711: ValueError
```

which is completely reasonable - the relationship just doesn't exist.


As for why `FieldDoesNotExist` vs `AttributeError`, I'm fine with `AttributeError` instead.

The `FieldDoesNotExist` exception came up when I tried to do what @john-westcott-iv originally suggested with deleting the attributes after we call the superclass `save()`. 

```python
def save(self, *args, **kwargs):
    # ...
    ret = super().save(*args, **kwargs)
    delattr(ImmutableModel, 'modified')
    return ret
```

...

```python
In [1]: l = ImmutableLogEntry(message="foo")

In [2]: l.save()

In [3]: l.modified
```

This throws a _crazy long_ stack trace, that ends with:
```
FieldDoesNotExist: ImmutableLogEntry has no field named 'modified'
```

So in this case `CommonModel` (the "other" base class) says "Sure, I have a modified field!" but the migration for the model definition says we don't have that field (since `ImmutableModel` nullifies it so Django's metaclass logic never turns it into a field).

`AttributeError` probably does make more sense coming from `__getattribute__()` - I'll change it to that.

I expected this to all be a pretty trivial change, doing it like @john-westcott-iv said, but it actually took me on a journey. In the end I came to the conclusion that the `__getattribute__()` override seems like the simplest way to say "hey, don't use these attributes on an `ImmutableModel` instance. If you have other ideas, I'm of course open to them.

Reply:
I've updated it to throw AttributeError now.